### PR TITLE
Move data fetcher ownership to the loops

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Advantages:
- Users can now customize fetching via loop customization.
- Each loop controls it's own fetcher logic.
- Removes clutter from the data connector.

Part of #11538 

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [n/a] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

cc @borda @justusschock @awaelchli @akihironitta @rohitgr7 @ninginthecloud @carmocca @ananthsub

### CODE REVIEW

File: tests/utilities/test_fetching.py [hunk 1]
```diff
@@ -251,7 +251,7 @@ def training_step(self, dataloader_iter, batch_idx):
 
         def training_epoch_end(self, *_):
             assert self.trainer.fit_loop.epoch_loop.batch_progress.current.ready == 33
-            assert self.trainer._data_connector.train_data_fetcher.fetched == 64
+            assert self.trainer.fit_loop._data_fetcher.fetched == 64
```
Comment (rohitgr7) [inline tests/utilities/test_fetching.py:254]: do we need similar tests for eval?

Comment (carmocca) [reply tests/utilities/test_fetching.py:254]: controlling the dataloader in the `step` is only supported for training

Comment (rohitgr7) [reply tests/utilities/test_fetching.py:254]: oh! correct, I didn't see that it uses just the `DataFetcher`

**Review by rohitgr7 — APPROVED** [review]
all green 👏

File: pytorch_lightning/loops/fit_loop.py [hunk 2]
```diff
@@ -325,8 +341,26 @@ def on_run_end(self) -> None:
...
+        rank_zero_warn(
+            "Found `dataloader_iter` argument in the `training_step`. Note that the support for "
+            "this signature is experimental and the behavior is subject to change."
+        )
+        return DataLoaderIterDataFetcher
+    elif os.getenv("PL_INTER_BATCH_PARALLELISM", "0") == "1":
+        if not isinstance(trainer.accelerator, GPUAccelerator):
+            raise MisconfigurationException("Inter batch parallelism is available only when using Nvidia GPUs.")
+        return InterBatchParallelDataFetcher
+    return DataFetcher
```
Comment (ananthsub) [inline pytorch_lightning/loops/fit_loop.py:366]: for this use case, do you recommend defining a utility function like this per loop? https://github.com/PyTorchLightning/pytorch-lightning/pull/11546/files

Comment (carmocca) [reply pytorch_lightning/loops/fit_loop.py:366]: Yes. Only that loop is meant to use it

File: pytorch_lightning/loops/fit_loop.py [hunk 3]
```diff
@@ -324,8 +340,26 @@ def on_run_end(self) -> None:
...
+            self._data_fetcher.teardown()
+            self._data_fetcher = None
         self.epoch_loop.teardown()
 
     def _should_accumulate(self) -> bool:
         """Whether the gradients should be accumulated."""
         return self.epoch_loop._should_accumulate()
+
+
+def _select_data_fetcher(trainer: "pl.Trainer") -> Type[AbstractDataFetcher]:
```
Comment (tchaton) [inline pytorch_lightning/loops/fit_loop.py:353]: Why not make this one a static method of the FitLoop?

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/loops/fit_loop.py (hunk 3 above)
Comment (carmocca) [reply pytorch_lightning/loops/fit_loop.py:353]: I'd say an independent function is generally preferred (simpler, no self call). But no strong argument

**Review by kaushikb11 — APPROVED** [review]
Great cleanup!
